rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -0,0 +1,391 @@
+```yaml
+# name: installTeleport
+---
+schemaVersion: '2.2'
+description: aws:runShellScript
+parameters:
+  token:
+    types: String
+    description: "(Required) The Teleport invite token to use when joining the cluster."
+mainSteps:
+- action: aws:downloadContent
+  name: downloadContent
+  inputs:
+    sourceType: "HTTP"
+    destinationPath: "/tmp/installTeleport.sh"
+    sourceInfo:
+      url: "https://teleportcluster.xyz/webapi/scripts/installer"
+- action: aws:runShellScript
+  name: runShellScript
+  inputs:
+    timeoutSeconds: '300'
+    runCommand:
+      - /bin/sh /tmp/installTeleport.sh "{{ token }}"
+```
+
+In order to run the new SSM document the AWS user will need IAM permissions to run
+SSM commands[3] for example:
+
+```js
+{
+    "Statement": [
+        {
+            "Action": "ssm:SendCommand",
+            "Effect": "Allow",
+            "Resource": [
+                # Allow running commands on all us-west-2 instances
+                "arn:aws:ssm:us-west-2:*:instance/*",
+                 # Allows running the installTeleport docuemnt on the allowed instances
+                "arn:aws:ssm:us-east-2:aws-account-ID:document/installTeleport"
+            ]
+        },
+		// "CreateDocument" and "GetDocument" permissions are required
+		// to automatically create the document
+        {
+            "Action": "ssm:CreateDocument",
+            "Effect": "Allow",
+            "Resource": [ "*" ]
+        },
+		{
+            "Action": "ssm:GetDocument",
+            "Effect": "Allow",
+            "Resource": [ "*" ]
+        }		
+    ]
+}
+```
+
+The machines being discovered will need to allow recieving `ec2messages` in
+order to recieve the SSM commands:
+
+```js 
+{
+	"Statement": [
+		{
+			"Action": "ec2messages:GetMessages"
+			"Effect": "Allow"
+		}
+	]
+}
+```
+
+On AWS, Amazon Linux and Ubuntu LTS (16.04, 18.04, 20.04) come with the SSM agent
+preinstalled[4].
+
+In order to allow nodes to create tokens for the purposes of sending invites to EC2
+instances a new system role will be added -- `RoleNodeDiscovery`, that will have
+permissions to create tokens.
+
+Each EC2 instance that is to be discovered will also require that they have an IAM
+role attached, in order to be able to send and recieve messages for the SSM agent.
+
+Example:
+
+```json
+{
+    "Statement": [
+        {
+            "Action": "ec2messages:*",
+            "Effect": "Allow",
+            "Resource": [
+                # Allow running commands on all us-west-2 instances
+                "*"
+            ]
+        }
+    ]
+}
+```
+
+
+## teleport.yaml generation
+
+The `teleport node configure` subcommand will be used to generate a
+new /etc/teleport.yaml file:
+```sh
+teleport node configure
+    --auth-server=auth-server.example.com [auth server that is being connected to]
+    --token="$1" # passed via parameter from SSM document
+    --labels=teleport.dev/origin=cloud
+	--node-name=$accountID-$instanceID
```

Comment:
Should probably update this command as we no longer set the node name and use labels instead?